### DISCUSSION TITLE
Evenly Distribute Space Dynamically

### DISCUSSION THREAD

**Discussion by waffle-lord** [AvaloniaUI/Avalonia#10786] [Q&A] [answered]
I was wondering how I could evenly space multiple controls. 

My idea is to have a vertical progressbar with ellipses on it to track task progression.

# Attempt with code-behind
I have something that looks the part, but relies entirely on code-behind making it seemingly impossible for me to bind properties to the controls.

I've read this, but it doesn't seem to provide any examples that make sense to me and all my attempts to use any of the methods to bind in code behind simply fail
https://docs.avaloniaui.net/docs/data-binding/binding-from-code

![image](https://user-images.githubusercontent.com/76401815/227644359-7b20f973-de8e-4762-941d-206d1e1d044d.png)

```cs
// I just picked a random event so I could make sure the Tasks property was set
public async void Control_AttachedToVisualTree(object sender, VisualTreeAttachmentEventArgs e)
{
    if (Tasks == null) return;

    var taskGrid = this.FindControl<Grid>("taskGrid");
    var taskProgressBar = this.FindControl<ProgressBar>("taskProgressBar");

    var rowDefs = string.Concat(Enumerable.Repeat("AUTO, *,", Tasks.Count));

    for (int i = 0; i < Tasks.Count; i++)
    {
        var taskItem = new ProgressableTaskItem()

        // How do you bind stuff in code behind?
        taskItem.TaskId = Tasks[i].Id;
        taskItem.TaskName = Tasks[i].Name;
        taskItem.IsRunning = Tasks[i].IsRunning;
        taskItem.IsCompleted = Tasks[i].IsCompleted;

        taskItem.SetValue(Grid.RowProperty, i * 2);

        taskGrid.Children.Add(taskItem);
    }

    // removes trailing ", *," 
    rowDefs = rowDefs.Remove(rowDefs.Length - 4);

    taskProgressBar.SetValue(Grid.RowSpanProperty, Tasks.Count * 2);
    taskGrid.RowDefinitions = new RowDefinitions(rowDefs);
}
```

The control is just a grid with a progress bar as the first child
```xaml
<Grid Name="taskGrid" Margin="20"
	  AttachedToVisualTree="Control_AttachedToVisualTree">

	<!-- progressbar values are inverted because I needed the progress to go top-to-bottom -->
	<!-- and vertical only seems to go bottom-to-top? -->
	<ProgressBar Name="taskProgressBar" Orientation="Vertical"
			     Value="25"
			     Background="DodgerBlue"
			     Foreground="Gainsboro"
			     HorizontalAlignment="Left"
			     VerticalAlignment="Stretch"
			     Margin="20 0"
			     />
</Grid>
```

# Attempt with Templates
I did also try to use a templates, but the items don't seem to be added to the grid itself and just get clumped at the top, even though the grid exists and has all the rows needed. Which is a shame, because it binds all the data perfectly, even though I still need some code-behind to setup the row definitions and row span  :(

Using a template would be ideal, but I can't seem to get the items as children of the grid, even though it is set as the item panel

![image](https://user-images.githubusercontent.com/76401815/227646769-fb97bbdf-4234-44c3-a366-efd20043c395.png)

```xaml
<ItemsControl Name="itemsControl" Items="{Binding Tasks, RelativeSource={RelativeSource
			  AncestorType=UserControl}}">
	<ItemsControl.ItemsPanel>
		<ItemsPanelTemplate>
			
			<Grid Margin="20" ShowGridLines="true"
				  AttachedToVisualTree="Control_AttachedToVisualTree">
				
				<ProgressBar Name="taskProgressBar" Orientation="Vertical"
						     AttachedToVisualTree="Control_AttachedToVisualTree"
						     Value="25"
						     Background="DodgerBlue"
						     Foreground="Gainsboro"
						     HorizontalAlignment="Left"
						     VerticalAlignment="Stretch"
						     Margin="20 0"
			     />
			</Grid>
		</ItemsPanelTemplate>
	</ItemsControl.ItemsPanel>
	<ItemsControl.ItemTemplate>
		<DataTemplate>
			<cc:ProgressableTaskItem TaskId="{Binding Id}"
				                                TaskName="{Binding Name}"
								IsRunning="{Binding IsRunning}"
								IsCompleted="{Binding IsComplted}"
								AttachedToVisualTree="Control_AttachedToVisualTree"
								/>
		</DataTemplate>
	</ItemsControl.ItemTemplate>
</ItemsControl>
```

I had to add the event to the grid, progress bar, and the template item since I need to update them in code-behind after the row definitions are setup.
```cs
public async void Control_AttachedToVisualTree(object sender, VisualTreeAttachmentEventArgs e)
{
    if (Tasks == null) return;

    if(sender is Grid taskGrid)
    {
        var rowDefs = string.Concat(Enumerable.Repeat("AUTO, *,", Tasks.Count));
        rowDefs = rowDefs.Remove(rowDefs.Length - 4);

        taskGrid.RowDefinitions = new RowDefinitions(rowDefs);
        return;
    }

    if(sender is ProgressBar progressBar)
    {
        if (Tasks == null) return;

        progressBar.SetValue(Grid.RowSpanProperty, Tasks.Count * 2);
        return;
    }

    if(sender is ProgressableTaskItem item)
    {
        var index = Tasks.IndexOf(Tasks.Where(x => x.Id == item.TaskId).First());

        item.SetValue(Grid.RowProperty, index * 2);
    }
}
```

So basically, I'm lost  :D

I feel like I'm fighting the API and just doing things the wrong way... but I'm having no luck finding information on how I should be tackling this setup.

Any help is greatly appreciated

P.S.
I have no idea why github keeps messing up the formatting of the xaml code sections. Sorry :(

**Comment by waffle-lord** [comment, marked as answer]
I managed to get something working with a custom panel implementation, but if there is a better way, I'd love to hear about it still.

**Comment by timunie** [comment]
I think a `UniformGrid` is what you are looking for. You may need number of rows to number of elements though.

**Reply by waffle-lord** [reply]
This is the control I couldn't remember, Thank you!

Sadly ...
I wasn't able to align the items in the uniform grid to the edges. They seem to always float, either to one side or in the middle depending on vertical alignment.

Since I need a progress bar to line up behind them, I can't have them floating around. I also can't seem to row span items, since I'd like the progress bar to be part of the item panel template.

So for now, I'm sticking with my custom panel

**Comment by lukaszossowski** [comment]
What are the actual requirements? Does the amount of Tasks vary?

I think I have a similarly created control that creates a row of radio buttons depending on the amout of them needed and binds each radio button to the value property. The creation of those radio buttons looks like this:

```
  private LHRadioPanel GenerateRadioButton(int id)
        {
            LHRadioPanel button = new LHRadioPanel();
            button.ID = id;
            button[!LHRadioPanel.ValueProperty] = this[!ValueProperty];
            button.Classes.Add("Button");
            return button;
        }
```
the LHRadioPanel is a custom radiobutton, but you can see here how binding is done in code behind.

**Reply by waffle-lord** [reply]
The amount of tasks can vary. The first and last items need to align to the edge of the panel and the middle items need to divide the space between them evenly. This is because I will have a progress bar behind them tracking overall progress through the tasks, and need the progress fill to line up with each item.

I did manage to get this working in a custom panel, I just don't know if that is the way I should do it.... it does work rather well though and allows me to continue to use a templated approach for bindings without all that messy event stuff I was doing originally.

I still don't really get how the binding in code behind stuff works tbh 🙃
but maybe there is just a limitation in my use case since I have an array and I'm trying to bind to a single element and not the entire array.

I appreciate the information though, maybe I'll take a second try at it that way just to see if I can better understand how the code behind bindings work.
